pkg/gardenlet/controller/federatedseed/network-policy/controller.go

```diff
@@ -0,0 +1,201 @@
+
+	GardenNamespaceSelector = labels.SelectorFromSet(labels.Set{
+		v1beta1constants.LabelRole: v1beta1constants.GardenNamespace})
+)
+
+// NewController instantiates a new controller.
+func NewController(ctx context.Context, seedClient kubernetes.Interface, seedDefaultNamespaceKubeInformer kubeinformers.SharedInformerFactory, seedKubeInformerFactory kubeinformers.SharedInformerFactory, seedLogger *logrus.Entry, recorder record.EventRecorder, seedName string) *Controller {
+	var (
+		endpointsInformer = seedDefaultNamespaceKubeInformer.Core().V1().Endpoints()
+		endpointsLister   = endpointsInformer.Lister()
+
+		networkPoliciesInformer = seedKubeInformerFactory.Networking().V1().NetworkPolicies()
+		networkPoliciesLister   = networkPoliciesInformer.Lister()
+
+		namespaceInformer = seedKubeInformerFactory.Core().V1().Namespaces()
+		namespaceLister   = namespaceInformer.Lister()
+		namespaceQueue    = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "namespace")
+	)
+
+	controller := &Controller{
+		log:                 seedLogger,
+		namespaceReconciler: newNamespaceReconciler(ctx, seedLogger, seedClient.Client(), endpointsLister, networkPoliciesLister, seedName),
+		recorder:            recorder,
+
+		endpointSynced:  endpointsInformer.Informer().HasSynced,
+		endpointsLister: endpointsLister,
+
+		networkPoliciesLister: networkPoliciesLister,
+		networkPoliciesSynced: networkPoliciesInformer.Informer().HasSynced,
+
+		namespaceSynced: namespaceInformer.Informer().HasSynced,
+		namespaceQueue:  namespaceQueue,
+		namespaceLister: namespaceLister,
+
+		workerCh: make(chan int),
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			endpoints, ok := obj.(*corev1.Endpoints)
+			if !ok {
+				return false
+			}
+			return endpoints.Name == "kubernetes"
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.endpointAdd,
+			UpdateFunc: controller.endpointUpdate,
+			DeleteFunc: controller.endpointDelete,
+		},
+	})
+
+	networkPoliciesInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			policy, ok := obj.(*networkingv1.NetworkPolicy)
+			if !ok {
+				return false
+			}
+			return policy.Name == helper.AllowToSeedAPIServer
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			UpdateFunc: controller.networkPoliciesUpdate,
+			DeleteFunc: controller.networkPoliciesDelete,
+		},
+	})
+
+	namespaceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			ns, ok := obj.(*corev1.Namespace)
+			if !ok {
+				return false
+			}
+			namespaceLables := labels.Set(ns.Labels)
+			if GardenNamespaceSelector.Matches(namespaceLables) {
+				return true
+			}
+
+			return ShootNamespaceSelector.Matches(namespaceLables)
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
```

Comment:
Shouldn't you also react on Update and Delete? Then we can probably get rid of this strange special handling for the garden namespace, see below?

Reply:
There should not be any need to react on those events - the only changes which might happen are not relative to the logic of the controller. We don't care that a namespace is deleted or it's state is set to terminating as we don't want to enqueue in this case.

Reply:
I agree, there is no use-case today. I was wondering whether it'd still would be better (from the pov of the design of this controller) to add them now in order to prevent future drawbacks. For example, this controller adds a new network policy to namespaces with a special label. Shouldn't it then also remove the network policy again when this label is removed from the namespace or when the namespace gets a deletion timestamp?
Yes, a delete handler is not necessary as there is nothing to be done, makes sense.

Reply:
You are right, for adding/removing labels

Reply:
Added the onUpdate function.